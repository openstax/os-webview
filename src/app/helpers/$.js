// FIX:  Delete as many of these as possible
const $ = {};

$.isTouchDevice = () => (
    ('ontouchstart' in window) ||
    (navigator.MaxTouchPoints > 0) ||
    (navigator.msMaxTouchPoints > 0)
 );

$.stringCompare = (a, b) => (a < b) ? -1 : +(a > b);

$.lowerCaseCompare = (a, b) => $.stringCompare(a.toLowerCase(), b.toLowerCase());

const tick = 1000 / 40;
const spaceForMenu = 59;
const targetStep = 200;
const targetTicks = 20;

$.scrollTo = (el, offset = 0) => {
    const rect = el.getBoundingClientRect();
    const offsetTop = rect.top - spaceForMenu - offset;
    const direction = Math.sign(offsetTop);
    let magnitude = Math.abs(offsetTop);
    const chosenStep = (targetStep + magnitude / targetTicks) / 2;

    const i = setInterval(() => {
        const step = (magnitude > chosenStep) ? chosenStep : magnitude;
        const scrollBody = document.documentElement.scrollTop || document.body.scrollTop;

        window.scrollTo(0, scrollBody + direction * step);
        magnitude -= step;
        if (magnitude <= 0) {
            clearInterval(i);
        }
    }, tick);
};

$.applyScrollFix = (view) => {
    let freezePosition = null;
    const setFreezePosition = () => {
        freezePosition = document.documentElement.scrollTop || document.body.scrollTop;
    };
    const handleMouseLeave = (e) => {
        if (!e.currentTarget.contains(e.relatedTarget)) {
            freezePosition = null;
        }
    };
    const handleWheelEvent = (e) => {
        const el = e.currentTarget;
        const delta = e.deltaY || e.wheelDelta / 4;

        el.scrollTop = el.scrollTop + delta;
        e.preventDefault();
    };
    const touchOutside = (el) => (
        (e) => {
            if (!el.contains(e.target)) {
                freezePosition = null;
            }
        }
    );
    let scrollStart;
    const setScrollStart = (e) => {
        scrollStart = e.currentTarget.scrollTop + e.touches[0].pageY;
    };
    const scroll = (e) => {
        e.currentTarget.scrollTop = scrollStart - event.touches[0].pageY;
        event.preventDefault();
    };

    for (const el of view.el.querySelectorAll('.mac-scroll')) {
        view.attachListenerTo(el, 'scroll', setFreezePosition);
        view.attachListenerTo(el, 'mouseleave', handleMouseLeave);
        view.attachListenerTo(el, 'mousewheel', handleWheelEvent);
        view.attachListenerTo(window, 'touchstart', touchOutside(el));
        if ($.isTouchDevice() && window.innerWidth < 769) {
            view.attachListenerTo(el, 'touchstart', setScrollStart);
            view.attachListenerTo(el, 'touchmove', scroll);
        }
    }
    view.attachListenerTo(window, 'scroll', () => {
        if (freezePosition !== null) {
            window.scrollTo(0, freezePosition);
        }
    });
};

$.hashClick = (event, options = {doHistory: true}) => {
    const node = event.delegateTarget;
    const destUrl = `${node.pathname}${node.hash}`;
    const targetEl = document.getElementById(node.hash.substr(1));

    $.scrollTo(targetEl);
    if (options.doHistory) {
        history.pushState({}, '', destUrl);
    }
    event.preventDefault();
};

const invalidEmailPatterns = [
    /@(aol|gmail|hotmail|yahoo).com/i
];

$.testInstitutionalEmail = (element) => {
    const ieValue = element.value;

    for (const re of invalidEmailPatterns) {
        if (re.test(ieValue)) {
            return false;
        }
    }

    return true;
};

$.htmlToText = (html) => {
    const temp = document.createElement('div');

    temp.innerHTML = html;
    return temp.textContent;
};

$.parseSearchString = (searchString) => {
    const result = {};

    searchString.substr(1).split('&').forEach((item) => {
        const [k, v] = item.split('=');

        (k in result) ? result[k].push(decodeURIComponent(v)) : result[k] = [decodeURIComponent(v)];
    });
    return result;
};

export default $;
