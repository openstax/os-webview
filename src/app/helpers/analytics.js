import settings from 'settings';
import linkHelper from '~/helpers/link';

const RELATIVE_TO_ROOT = /^\//;
const SETUP_GTM = Symbol();
const SETUP_GA = Symbol();

class Analytics {

    constructor() {
        this.start();
    }

    start() {
        this[SETUP_GA]();
        this[SETUP_GTM]();
    }

    send(fields) {
        if (this.tracking) {
            SystemJS.import('ga').then(() => {
                window.ga('send', fields);
            });
        }
    }

    sendPageview(page) {
        let frag = page || location.pathname;

        if (!(RELATIVE_TO_ROOT).test(frag)) {
            frag = `/${frag}`;
        }

        this.send({
            hitType: 'pageview',
            page: frag
        });
    }

    sendEvent(fields) {
        this.send(Object.assign(
            {hitType: 'event'},
            fields
        ));
    }

    record(href) {
        if (linkHelper.isPDF(href)) {
            this.sendEvent({
                eventCategory: 'PDF',
                eventAction: 'download',
                eventLabel: href
            });
        }

        if (linkHelper.isExternal(href)) {
            this.sendEvent({
                eventCategory: 'External',
                eventAction: 'open',
                eventLabel: href
            });
        }
    }

    [SETUP_GTM]() {
        (function (w, d, s, l, i) {
            // Disable ESLint rules since we're copying Google's script
            /* eslint max-params: 0 */
            /* eslint eqeqeq: 0 */
            /* eslint prefer-template: 0 */

            w[l] = w[l] || [];
            w[l].push({
                'gtm.start': new Date().getTime(),
                event: 'gtm.js'
            });

            const f = d.getElementsByTagName(s)[0];
            const j = d.createElement(s);
            const dl = l !== 'dataLayer' ? '&l=' + l : '';

            j.async = true;
            j.src = '//www.googletagmanager.com/gtm.js?id=' + i + dl;
            f.parentNode.insertBefore(j, f);
        })(window, document, 'script', 'dataLayer', settings.tagManagerID);
    }

    [SETUP_GA]() {
        if (typeof window.ga !== 'function') {
            window.GoogleAnalyticsObject = 'ga';
            window.ga = {
                q: [['create', settings.analyticsID, 'auto']],
                l: Date.now()
            };
        } else {
            window.ga('create', settings.analyticsID, 'auto');
        }

        document.addEventListener('submit', (e) => {
            if (typeof e.target !== 'object' || typeof e.target.action !== 'string') {
                return;
            }

            if (e.target.action.indexOf('https://www.salesforce.com/') === 0) {
                const formData = new FormData(e.target);

                this.sendEvent({
                    eventCategory: 'Salesforce',
                    eventAction: 'submit',
                    eventLabel: formData.get('lead_source')
                });
            }
        });
    }

}

const analytics = new Analytics();

export default analytics;
