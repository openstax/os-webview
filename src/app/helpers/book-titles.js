let comingSoon = [
    'Astronomy',
    'Elementary Algebra',
    'Intermediate Algebra',
    'Microbiology',
    {
        text: 'University Physics',
        value: 'University Physics (Calc)'
    }
];

export let published = [
    'Algebra and Trigonometry',
    'American Government',
    'Anatomy & Physiology',
    'Biology',
    'Calculus',
    'Chemistry',
    {
        text: 'Chemistry: Atoms First',
        value: 'Chem: Atoms First'
    },
    'College Algebra',
    {
        text: 'College Physics',
        value: 'College Physics (Algebra)'
    },
    {
        text: 'College Physics for AP® Courses',
        value: 'AP Physics'
    },
    {
        text: 'Concepts of Biology',
        value: 'Concepts of Bio (non-majors)'
    },
    'Introduction to Sociology',
    {
        text: 'Introduction to Sociology 2e',
        value: 'Introduction to Sociology'
    },
    'Introductory Statistics',
    {
        text: 'Prealgebra',
        value: 'PreAlgebra'
    },
    {
        text: 'Precalculus',
        value: 'Precalc'
    },
    {
        text: 'Principles of Economics',
        value: 'Economics'
    },
    {
        text: 'Principles of Macroeconomics',
        value: 'Macro Econ'
    },
    {
        text: 'Principles of Macroeconomics for AP® Courses',
        value: 'AP Macro Econ'
    },
    {
        text: 'Principles of Microeconomics',
        value: 'Micro Econ'
    },
    {
        text: 'Principles of Microeconomics for AP® Courses',
        value: 'AP Micro Econ'
    },
    'Psychology',
    {
        text: 'U.S. History',
        value: 'US History'
    }
];

let bookTitles = [...comingSoon, ...published].sort((a, b) => {
    let aText = a.text || a,
        bText = b.text || b;

    return aText < bText ? -1 : aText !== bText;
});

export default bookTitles;
