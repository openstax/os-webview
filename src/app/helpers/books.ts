export type Book = {
    book_state: string;
    cover_url: string;
    salesforce_abbreviation: string;
    salesforce_name: string;
    slug: string;
    subjects: string[];
    title: string;
};

const statesToInclude = ['live', 'new_edition_available', 'coming_soon'];

export const filterLiveBooks = (entry: Book) =>
    statesToInclude.includes(entry.book_state);

function BookToSalesforceBook(book: Book) {
    return {
        text: book.salesforce_name,
        value: book.salesforce_abbreviation,
        comingSoon: book.book_state === 'coming_soon',
        subjects: book.subjects,
        coverUrl: book.cover_url
    };
}

export type SalesforceBook = ReturnType<typeof BookToSalesforceBook>;

export function salesforceTitles(books: Book[]) {
    const seenTitles: {[key: string]: boolean} = {};

    return books
        .filter(filterLiveBooks)
        .filter((book) => {
            const abbrev = book.salesforce_abbreviation;
            const seen = abbrev in seenTitles;

            seenTitles[abbrev] = true;
            return abbrev && !seen;
        })
        .map(BookToSalesforceBook)
        .sort((a, b) => (a.text < b.text ? -1 : 1));
}

export function subjects(sfTitles: Book[]) {
    return sfTitles
        .reduce<string[]>((a, b) => a.concat(b.subjects || []), [])
        .reduce<string[]>((a, b) => (a.includes(b) ? a : a.concat(b)), []);
}
