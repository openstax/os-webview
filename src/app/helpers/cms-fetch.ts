import retry from '~/helpers/retry';
import urlFromSlug from './url-from-slug';

export default async function cmsFetch(path) {
    const url = path.replace(/[^?]+/, urlFromSlug);

    try {
        return (await retry(() => fetch(url))).json();
    } catch (err) {
        return Promise.reject(new Error(`Failed to fetch ${path}: ${err}`));
    }
}

export async function cmsPost(path, payload, method='POST') {
    const url = path.replace(/[^?]+/, urlFromSlug);
    const params = new window.URLSearchParams(payload);
    const qs = method.toLowerCase() === 'delete' ? `?${params}` : '';

    try {
        return (await fetch(`${url}${qs}`, {
            method,
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(payload)
        })).json();
    } catch (e) {
        return Promise.reject(new Error(`Failed to ${method} ${url}${qs}`));
    }
}
