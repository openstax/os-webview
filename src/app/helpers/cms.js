import settings from 'settings';

const jsonToQueryString = (json) =>
    Object.keys(json).map((key) =>
        `${encodeURIComponent(key)}=${encodeURIComponent(json[key])}`
    ).join('&');

const cms = {
    query: (searchOptions) => {
        const url = `${settings.apiOrigin}/api/v1/pages/?format=json&${jsonToQueryString(searchOptions)}`;

        return new Promise((resolve) => fetch(url)
        .then((response) => response.json())
        .then(resolve));
    }
};

export default cms;
