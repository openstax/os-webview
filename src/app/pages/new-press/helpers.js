import React from 'react';
import Byline from '~/components/byline/byline';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faExternalLinkAlt} from '@fortawesome/free-solid-svg-icons/faExternalLinkAlt';

export function ContentBlock({title, children}) {
    return (
        <div className='content-block'>
            <h2>{title}</h2>
            {children}
        </div>
    );
}

export function asDate(dateStr) {
    const [year, month, day] = dateStr.split('-');

    return new Date(year, month - 1, day);
}

export function convertedDate(dateStr) {
    return asDate(dateStr).toLocaleString('en-US', {
        month: 'short',
        day: 'numeric',
        year: 'numeric'
    });
}

function OptionalExcerpt({excerpt, url}) {
    return (
        excerpt && (
            <div className='excerpt'>
                {excerpt}â€¦ <a href={url}>Continue reading</a>
            </div>
        )
    );
}

export function PressExcerpt({
    iconUrl,
    author,
    date,
    source,
    url,
    headline,
    excerpt
}) {
    const classList = ['press-excerpt'];

    if (iconUrl) {
        classList.push('with-icon');
    }

    return (
        <div className={classList.join(' ')}>
            {iconUrl && <img src={iconUrl} alt='' />}
            <Byline author={author} date={date} source={source} />
            <div className='headline'>
                <a href={url}>
                    {headline}{' '}
                    {source && <FontAwesomeIcon icon={faExternalLinkAlt} />}
                </a>
            </div>
            <OptionalExcerpt excerpt={excerpt} url={url} />
        </div>
    );
}
