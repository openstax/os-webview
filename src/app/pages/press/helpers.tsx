import React from 'react';
import Byline from '~/components/byline/byline';
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome';
import {faExternalLinkAlt} from '@fortawesome/free-solid-svg-icons/faExternalLinkAlt';

export function ContentBlock({title, children}: React.PropsWithChildren<{title: string}>) {
    return (
        <div className='content-block'>
            <h2>{title}</h2>
            {children}
        </div>
    );
}

export function asDate(dateStr: string) {
    const [year, month, day] = dateStr.split('-');

    return new Date(+year, +month - 1, +day);
}

export function convertedDate(dateStr: string) {
    return asDate(dateStr).toLocaleString('en-US', {
        month: 'short',
        day: 'numeric',
        year: 'numeric'
    });
}

function OptionalExcerpt({excerpt, url}: {excerpt?: string; url: string}) {
    return (
        excerpt && (
            <div className='excerpt'>
                {excerpt}â€¦ <a href={url}>Continue reading</a>
            </div>
        )
    );
}

export function PressExcerpt({
    iconUrl,
    author,
    date,
    source,
    url,
    headline,
    excerpt
}: Parameters<typeof Byline>[0] & {
    iconUrl?: string;
    url: string;
    headline: string;
    excerpt?: string;
}) {
    const classList = ['press-excerpt'];

    if (iconUrl) {
        classList.push('with-icon');
    }

    return (
        <div className={classList.join(' ')}>
            {iconUrl && <img src={iconUrl} alt='' />}
            <Byline author={author} date={date} source={source} />
            <div className='headline'>
                <a href={url}>
                    {headline}{' '}
                    {source && <FontAwesomeIcon icon={faExternalLinkAlt} />}
                </a>
            </div>
            <OptionalExcerpt excerpt={excerpt} url={url} />
        </div>
    );
}
