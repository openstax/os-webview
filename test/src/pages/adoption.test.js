import '../../helpers/fetch-mocker';
import AdoptionForm from '~/pages/adoption/adoption';
import instanceReady from '../../helpers/instance-ready';

describe('AdoptionForm', () => {
    const {instance:p, ready} = instanceReady(AdoptionForm);

    it('matches snapshot', () =>
        ready.then(() => {
            console.log('To update snapshot: node_modules/.bin/jest --updateSnapshot --testNamePattern=AdoptionForm');
            expect(p.el.innerHTML).toMatchSnapshot();
        })
    );
});
