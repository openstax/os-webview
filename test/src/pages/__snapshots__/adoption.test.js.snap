// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdoptionForm matches snapshot 1`] = `"<div data-region=\\"header\\"><div class=\\"form-header\\"></div></div><img class=\\"strips\\" src=\\"/images/components/strips.svg\\" height=\\"10\\" alt=\\"\\" role=\\"presentation\\"><div data-region=\\"role-selector\\" class=\\"text-content\\"><div class=\\"role-selector\\"><form data-region=\\"selector\\"><label class=\\"form-select\\"><div class=\\"control-group\\"><label class=\\"field-label\\" data-html=\\"labelHtml\\"></label><label class=\\"field-label\\"> I am a</label><span class=\\"invalid-message\\"></span><div class=\\"proxy-select select with-arrow\\" tabindex=\\"0\\"><span class=\\"item none\\">Please select one</span><ul class=\\"options\\"><li class=\\"option\\" data-value=\\"Student\\">Student</li><li class=\\"option\\" data-value=\\"Faculty\\">Instructor</li><li class=\\"option\\" data-value=\\"Home School Teacher\\">Homeschool Instructor</li><li class=\\"option\\" data-value=\\"Administrator\\">Administrator</li><li class=\\"option\\" data-value=\\"Librarian\\">Librarian</li><li class=\\"option\\" data-value=\\"Instructional Designer\\">Instructional Designer</li><li class=\\"option\\" data-value=\\"Adjunct Faculty\\">Adjunct Faculty</li><li class=\\"option\\" data-value=\\"Other\\">Other</li></ul></div><select placeholder=\\"Please select one\\"><option value=\\"Student\\">Student</option><option value=\\"Faculty\\">Instructor</option><option value=\\"Home School Teacher\\">Homeschool Instructor</option><option value=\\"Administrator\\">Administrator</option><option value=\\"Librarian\\">Librarian</option><option value=\\"Instructional Designer\\">Instructional Designer</option><option value=\\"Adjunct Faculty\\">Adjunct Faculty</option><option value=\\"Other\\">Other</option></select></div></label></form><div data-content=\\"0\\"><div class=\\"student-form\\"><div class=\\"text-content\\">Students don't need to fill out any forms to use our books. Access them free now!<div class=\\"cta\\"><button class=\\"btn\\" tabindex=\\"0\\">Go back</button></div></div></div></div><div hidden=\\"\\" data-content=\\"1\\"><div class=\\"multi-page-form\\"><iframe name=\\"form-response\\" id=\\"form-response\\" class=\\"hidden\\" src=\\"\\" width=\\"0\\" height=\\"0\\" tabindex=\\"-1\\"></iframe><form accept-charset=\\"UTF-8\\" class=\\"form\\" target=\\"form-response\\" method=\\"post\\" action=\\"https://test.salesforce.com/servlet/servlet.WebToLead?encoding=UTF-8\\"><div data-page=\\"0\\"><div class=\\"page-1\\"><div class=\\"hidden-fields\\"><input name=\\"utf8\\" type=\\"hidden\\" value=\\"âœ“\\"><input type=\\"hidden\\" name=\\"Application_Source__c\\" value=\\"OS Web/Accounts\\"><input type=\\"hidden\\" name=\\"oid\\" value=\\"00D0t0000000zUN\\"><input type=\\"hidden\\" name=\\"lead_source\\" value=\\"Adoption Form\\"><input type=\\"hidden\\" name=\\"Role__c\\" value=\\"Student\\"><input type=\\"hidden\\" name=\\"Subject__c\\"><input type=\\"hidden\\" name=\\"Adoption_Status__c\\"><input type=\\"hidden\\" name=\\"Number_of_Students__c\\"><input type=\\"hidden\\" name=\\"First__c\\"></div><div><plug-in data-id=\\"popup\\"></plug-in><plug-in data-id=\\"firstName\\"><label class=\\"form-input\\"><div class=\\"control-group\\"><label class=\\"field-label\\">First name</label><input name=\\"first_name\\" type=\\"text\\" value=\\"\\" required=\\"\\" autocomplete=\\"given-name\\" class=\\"\\"><div class=\\"invalid-message\\"></div></div></label></plug-in><plug-in data-id=\\"lastName\\"><label class=\\"form-input\\"><div class=\\"control-group\\"><label class=\\"field-label\\">Last name</label><input name=\\"last_name\\" type=\\"text\\" value=\\"\\" required=\\"\\" autocomplete=\\"family-name\\" class=\\"\\"><div class=\\"invalid-message\\"></div></div></label></plug-in><plug-in data-id=\\"email\\"><label class=\\"form-input\\"><div class=\\"control-group\\"><label class=\\"field-label\\">Email address</label><input name=\\"email\\" type=\\"email\\" value=\\"\\" required=\\"\\" autocomplete=\\"email\\" class=\\"\\"><div class=\\"invalid-message\\"></div></div></label></plug-in><plug-in data-id=\\"phone\\"><label class=\\"form-input\\"><div class=\\"control-group\\"><label class=\\"field-label\\">Phone number</label><input name=\\"phone\\" type=\\"text\\" value=\\"\\" required=\\"\\" autocomplete=\\"tel-national\\" class=\\"\\"><div class=\\"invalid-message\\"></div></div></label></plug-in><plug-in data-id=\\"school\\"><label class=\\"form-input\\"><div class=\\"control-group\\"><label class=\\"field-label\\">School name</label><input name=\\"company\\" type=\\"text\\" value=\\"\\" required=\\"\\" autocomplete=\\"organization\\" class=\\"\\"><div class=\\"invalid-message\\"></div></div></label></plug-in><plug-in data-id=\\"schoolUrl\\"></plug-in></div></div></div><div data-page=\\"1\\" hidden=\\"\\"><div class=\\"page-2\\"><div class=\\"book-selector\\"><div><h2 class=\\"prompt\\">Which textbook(s) are you currently using?</h2><div class=\\"hint\\">Select all that apply</div></div></div><div class=\\"how-using\\"></div></div></div><div data-page=\\"2\\" hidden=\\"\\"><div class=\\"technology-selector\\"><h2 class=\\"prompt\\">Tell us about the technology you use.</h2><div class=\\"hint\\">Check all that apply (optional)</div><div class=\\"two-columns\\" data-region=\\"checkboxes\\"></div></div></div><div class=\\"button-row\\"><button type=\\"button\\" class=\\"secondary back\\" hidden=\\"\\">Back</button><button type=\\"submit\\" class=\\"primary\\" hidden=\\"\\">Submit</button><button type=\\"button\\" class=\\"primary next\\">Next</button></div></form></div></div></div></div>"`;
